### PR TITLE
Bug 1321798 - Rewrite autoclassification frontend for better sheriff workflow

### PR DESCRIPTION
**Pull request by jgraham**
Please ignore the first two commits in this review (i.e. any changes to non-js files); I need to rebase and land that work. But the js changes are rather large and it's way past time to get some review on them, so I think it's worth starting this process anyway.

---
This change is [Reviewable](https://reviewable.io/reviews/mozilla/treeherder/2089)

### CODE REVIEW

**Comment by jgraham** [comment]
Ah, also the fourth commit.

File: ui/plugins/auto_classification/controller.js [hunk 1]
```diff
@@ -69,891 +40,1104 @@ treeherder.factory('thStringOverlap', function() {
...
+                return;
             }
-            if (ui.best) {
-                ui.best.isBest = true;
+            if (id !== $scope.selectedOption.id) {
+                $scope.selectedOption.id = id;
+                $scope.optionChanged();
             }
+            if (option === "=") {
+                $("#" + line.id + "-manual-bug").focus();
```
Comment (wlach) [inline ui/plugins/auto_classification/controller.js]: Kind of a bummer to be using raw jquery here, but I assume you couldn't find an alternative.

File: ui/plugins/auto_classification/controller.js [hunk 2]
```diff
@@ -69,891 +40,1104 @@ treeherder.factory('thStringOverlap', function() {
...
-                                $scope.failureLines = buildFailureLineOptions(data);
-                            })
-                            .finally(function() {
-                                thTabs.tabs.autoClassification.is_loading = false;
-                                $scope.triedLoad = true;
-                            });
-                    }
+            var resources = {
+                "matchers": ThMatcherModel.by_id(),
+
```
Comment (wlach) [inline ui/plugins/auto_classification/controller.js]: Is this useful?

Comment (wlach) [inline ui/plugins/auto_classification/controller.js]: Nit: Do you actually need/want to create a temporary variable here?

File: ui/plugins/auto_classification/controller.js [hunk 3]
```diff
@@ -69,891 +40,1104 @@ treeherder.factory('thStringOverlap', function() {
...
-                                  function(matcher) {
-                                      matchersById[matcher.id] = matcher;
-                                  });
-                        return matchersById;
+        /**
+         * (Re)build all the panel contents with fresh data
+         */
+        function build() {
+            log.debug("build", ctrl);
+            log.debug("Status", ctrl.loadStatus);
```
Comment (wlach) [inline ui/plugins/auto_classification/controller.js]: Did you want to leave this in?

File: ui/plugins/auto_classification/controller.js [hunk 4]
```diff
@@ -69,891 +40,1104 @@ treeherder.factory('thStringOverlap', function() {
...
-                    scores.set(bug.id, score);
-                    return scores;
-                }, new Map());
+        /**
+         * Get the initial default option
+         * @param {Object[]} options - List of line-specific options
+         * @param {Object[]} extraOptions - List of line-independent options
+         * @param {Object[]} prevLine - Line before the current one in the log
+         */
+        function getDefaultOption(options, extraOptions, prevLine) {
```
Comment (camd) [inline ui/plugins/auto_classification/controller.js]: This function would really benefit from more comments.  You have some comments here, which helps, but the logic is pretty complex, so it'd be MUCH easier to review and understand later if you put in a play-by-play to some degree.

Something along the lines of: "Now we have determined that X is true, so we need to find out if we should Y or Z."  I may be the only one, but comments like that help me with a decision function such as this.  :)

**Comment by jgraham** [comment]
Review status: 0 of 49 files reviewed at latest revision, 17 unresolved discussions.

---

*[ui/index.html, line 141 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RJEp2aFrZ8xX3k-r12-141:-KccAPlDQ5n39BKXrW98:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/index.html#L141)):*
<details><summary><i>Previously, camd wrote…</i></summary><blockquote>

won't this allow the config file to be added to the minified js?  I don't think we want that, do we?  :)
</blockquote></details>

Done.

---

*[ui/js/controllers/bugfiler.js, line 10 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RPxfxjnUCWQHI2-r12-10:-KccAQxrIvd-4Kk0KDkq:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/js/controllers/bugfiler.js#L10)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary><blockquote>

It seems like there is a bunch of cleanup in here which isn't strictly related to your feature work. I'd prefer if this was put into another PR.
</blockquote></details>

Done.

---

*[ui/js/controllers/main.js, line 344 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RJEp2aFrZ8xX3l-r12-344:-KcPp8ZU0QEswpRmIVyg:b-4lri5w) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/js/controllers/main.js#L344)):*
<details><summary><i>Previously, camd wrote…</i></summary><blockquote>

this would be good to send out a note that these shortcuts will be changing.  I don't have a feel for how many people may use ``j`` and ``k`` this way, but it seems ripe for breaking someone's workflow...  :)  Doesn't mean we shouldn't do it.  Pretty unnecessary to have 2 keys to do the same thing...
</blockquote></details>

Sure, I have mentioned this to the sheriffs several times,  but I can mention it again with the inevitable launch email.

---

*[ui/js/controllers/main.js, line 525 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RJEp2aFrZ8xX3l-r12-525:-Kcc7q1uWktQcVmhJkCY:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/js/controllers/main.js#L525)):*
<details><summary><i>Previously, camd wrote…</i></summary><blockquote>

would you go ahead and comment the rest of these?  It will save someone some time down the line... :)
</blockquote></details>

Done.

---

*[ui/js/models/matcher.js, line 36 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RKg-BDFn85PESz-r12-36:-Kcc7tDop36zZqchD_ir:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/js/models/matcher.js#L36)):*
<details><summary><i>Previously, camd wrote…</i></summary><blockquote>

you don't actually need the parens around ``response`` and ``elem`` since it's just one param each
</blockquote></details>

Done.

---

*[ui/js/models/text_log_errors.js, line 28 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RKg-BDFn85PET7-r12-28:-Kcc7vONAt0e6Yhbb8yD:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/js/models/text_log_errors.js#L28)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary><blockquote>

Couldn't you just use `map` here?
</blockquote></details>

Done.

---

*[ui/plugins/tabs.js, line 8 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RSAS3uf1TusH_0-r12-8:-KccA01V85raQw3ooIZL:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/plugins/tabs.js#L8)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary><blockquote>

Likewise, could you put these changes in a separate PR?
</blockquote></details>

Done.

---

*[ui/plugins/auto_classification/controller.js, line 71 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RKg-BDFn85PET1-r12-71:-KccA3iM6a_fbbo8y6XM:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/plugins/auto_classification/controller.js#L71)):*
<details><summary><i>Previously, camd wrote…</i></summary><blockquote>

You could just ``return`` here since you don't use the var anywhere.
</blockquote></details>

Done.

---

*[ui/plugins/auto_classification/controller.js, line 117 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RKg-BDFn85PET1-r12-117:-KccA5O5yxq624_OMuwD:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/plugins/auto_classification/controller.js#L117)):*
<details><summary><i>Previously, camd wrote…</i></summary><blockquote>

Looks like a few of these bindings are never used.  Could they be removed?
</blockquote></details>

Done.

---

*[ui/plugins/auto_classification/controller.js, line 167 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RKg-BDFn85PET1-r12-167:-KccA7ae93jZss8pyzq-:b-exae48) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/plugins/auto_classification/controller.js#L167)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary><blockquote>

If you're going to parametrize the bugzilla link, may as well use that value here too.
</blockquote></details>

Done in the other PR; I'll rebase on to that when it lands.

---

*[ui/plugins/auto_classification/controller.js, line 442 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RKg-BDFn85PET1-r12-442:-KccAS2KfZxfEvECkvL2:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/plugins/auto_classification/controller.js#L442)):*
<details><summary><i>Previously, camd wrote…</i></summary><blockquote>

typo: "of if" sb "or if"
</blockquote></details>

Done.

---

*[ui/plugins/auto_classification/controller.js, line 496 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RKg-BDFn85PET1-r12-496:-KccAEVygZPZi0Ta85-y:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/plugins/auto_classification/controller.js#L496)):*
<details><summary><i>Previously, camd wrote…</i></summary><blockquote>

This function would really benefit from more comments.  You have some comments here, which helps, but the logic is pretty complex, so it'd be MUCH easier to review and understand later if you put in a play-by-play to some degree.

Something along the lines of: "Now we have determined that X is true, so we need to find out if we should Y or Z."  I may be the only one, but comments like that help me with a decision function such as this.  :)
</blockquote></details>

Done.

---

*[ui/plugins/auto_classification/controller.js, line 507 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RKg-BDFn85PET1-r12-507:-KccAIZNAH5wk923-PAs:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/plugins/auto_classification/controller.js#L507)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary><blockquote>

It's not that big a deal, but you could shorten this to:

```js
var parts = line.split(" | ", 3);
return (parts.length === 3) parts[1] : null;
```
</blockquote></details>

Done.

---

*[ui/plugins/auto_classification/controller.js, line 606 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RKg-BDFn85PET1-r12-606:-KcPukRA0Q1Bt0DXWyas:b-9xqpnd) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/plugins/auto_classification/controller.js#L606)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary><blockquote>

Kind of a bummer to be using raw jquery here, but I assume you couldn't find an alternative.
</blockquote></details>

I don't really know how to programmatically focus an element using angular. Unless it's easy it doesn't seem worthwhile.

---

*[ui/plugins/auto_classification/controller.js, line 837 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RKg-BDFn85PET1-r12-837:-KccAKNgssp3SshqdYSA:bss4a2b) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/plugins/auto_classification/controller.js#L837)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary><blockquote>

Did you want to leave this in?
</blockquote></details>

Removed.

---

*[ui/plugins/auto_classification/controller.js, line 907 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RKg-BDFn85PET1-r12-907:-KccAO9Q80df_fpB9W8h:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/plugins/auto_classification/controller.js#L907)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary><blockquote>

Nit: Do you actually need/want to create a temporary variable here?
</blockquote></details>

Done.

---

*[Gruntfile.js, line 167 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RJEp2aFrZ8xX3d-r12-167:-KcPoQe21ZRj_a1PbiJ6:b-n7o3za) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/Gruntfile.js#L167)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary><blockquote>

Was this change intentional?
</blockquote></details>

Well yes, but I can revert if needed. It just makes it easier to debug the minified code by adding whitespace. Presumably there is some small cost in load time.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/2089)*

File: tests/autoclassify/test_classify_failures.py [hunk 5]
```diff
@@ -37,45 +32,54 @@ def do_autoclassify(job, test_failure_lines, matchers, status="testfailed"):
...
-    expected_classified = test_failure_lines[:2]
-    expected_unclassified = test_failure_lines[2:]
+    expected_classified = test_error_lines[:2], test_failure_lines[:2]
+    expected_unclassified = test_error_lines[2:], test_failure_lines[2:]
 
-    for actual, expected in zip(expected_classified, classified_failures):
-        assert [item.id for item in actual.classified_failures.all()] == [expected.id]
+    for (error_line, failure_line), expected in zip(zip(*expected_classified),
+                                                    classified_failures):
+        assert [item.id for item in error_line.classified_failures.all()] == [expected.id]
```
Comment (wlach) [inline tests/autoclassify/test_classify_failures.py]: Not that it matters hugely, but I would do:

```py
error_line.classified_failures.values_list('id', flat=True) == [expected.id]
```

File: tests/autoclassify/test_classify_failures.py [hunk 6]
```diff
@@ -37,45 +32,54 @@ def do_autoclassify(job, test_failure_lines, matchers, status="testfailed"):
         item.refresh_from_db()
 
 
-def test_classify_test_failure(failure_lines, classified_failures,
+def test_classify_test_failure(text_log_errors_failure_lines,
+                               classified_failures,
                                test_job_2):
```
Comment (wlach) [inline tests/autoclassify/test_classify_failures.py]: Could you add a short comment stating what cases this test is supposed to be testing? Some of the other test cases in this file are obvious from the name, but`test_classify_test_failure` is pretty ambiguous...

File: tests/autoclassify/test_classify_failures.py [hunk 7]
```diff
@@ -37,45 +32,54 @@ def do_autoclassify(job, test_failure_lines, matchers, status="testfailed"):
...
-def test_no_autoclassify_job_success(failure_lines, classified_failures, test_job_2):
-    test_failure_lines = create_failure_lines(test_job_2,
-                                              [(test_line, {}),
-                                               (test_line, {"subtest": "subtest2"}),
-                                               (test_line, {"status": "TIMEOUT"}),
-                                               (test_line, {"expected": "ERROR"}),
-                                               (test_line, {"message": "message2"})])
+def test_no_autoclassify_job_success(text_log_errors_failure_lines,
+                                     classified_failures,
+                                     test_job_2):
```
Comment (wlach) [inline tests/autoclassify/test_classify_failures.py]: Ditto, a short comment describing what's being tested here would be helpful.

File: tests/ui/unit/controllers/autoclassify.tests.js [hunk 8]
```diff
@@ -1,104 +1,104 @@
 'use strict';
 
-describe('AutoClassify panel', function() {
+// describe('AutoClassify panel', function() {
```
Comment (wlach) [inline tests/ui/unit/controllers/autoclassify.tests.js]: I'll leave it up to you on whether it is worth refactoring this to reflect the current state of the code. But if you're just going to comment this out, I would rather remove the file.

File: treeherder/model/migrations/0055_autoclassify_status.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    operations = [migrations.RunSQL(
+        sql="""
+BEGIN;
+--
+-- Add field autoclassify_status to job
+--
+SET FOREIGN_KEY_CHECKS=0;
+ALTER TABLE `job` ADD COLUMN `autoclassify_status` integer DEFAULT 0 NOT NULL;
+ALTER TABLE `job` ALTER COLUMN `autoclassify_status` DROP DEFAULT;
+COMMIT;
```
Comment (wlach) [inline treeherder/model/migrations/0055_autoclassify_status.py]: I am curious if the manual SQL for this migration has any advantages over the automatically generated sql by Django? I know (boy do I know) that sometimes Django can be inefficient at doing these things, but I don't see how that would be the case here.

I'd prefer to use the default behaviour where possible, as it would make an eventual migration to another database easier (at the very least there would be less to audit).

File: treeherder/model/models.py [hunk 10]
```diff
@@ -1450,15 +1497,120 @@ class TextLogError(models.Model):
...
+    failure_line = models.OneToOneField(FailureLine,
+                                        related_name="text_log_error",
+                                        null=True)
+    # Note that the case of best_classification = None and best_is_verified = True
+    # has the special semantic that the line is ignored and should not be considered
+    # for future autoclassifications.
+    best_classification = models.ForeignKey("ClassifiedFailure",
+                                            related_name="best_for_errors",
+                                            null=True,
+                                            db_index=True,
```
Comment (wlach) [inline treeherder/model/models.py]: db_index is not necessary here, as django automatically indexes by column where a foreign key is specified.

File: treeherder/webapp/api/text_log_error.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    def _update(self, data, user, many=True):
+        ids = []
+        error_line_ids = set()
+        classification_ids = set()
+        bug_number_classifications = {}
+
+        for item in data:
+            line_id = int(item.get("id"))
+            if line_id is None:
+                return "No text log error id provided", HTTP_400_BAD_REQUEST
```
Comment (wlach) [inline treeherder/webapp/api/text_log_error.py]: if `item.get(..)` is None, casting it to an int is going to throw an exception.

File: treeherder/model/models.py [hunk 12]
```diff
@@ -1441,6 +1471,23 @@ class Meta:
                            'finished_line_number')
 
 
+class TextLogErrorManager(models.Manager):
```
Comment (wlach) [inline treeherder/model/models.py]: Could you add one-line comments for each of the methods in here describing what they do?

File: treeherder/autoclassify/autoclassify.py [hunk 13]
```diff
@@ -20,57 +26,85 @@ def match_errors(job):
...
+    if not unmatched_errors:
+        logger.info("Skipping autoclassify of job %i because it has no unmatched errors" % job.id)
+        return
 
-def find_matches(unmatched_failures):
+    try:
+        matches, all_matched = find_matches(unmatched_errors)
+        update_db(job, matches, all_matched)
+    except:
+        logger.debug("Autoclassification of job %s failed" % job.id)
```
Comment (wlach) [inline treeherder/autoclassify/autoclassify.py]: Shouldn't this be log level error?

File: treeherder/autoclassify/detect_intermittents.py [hunk 14]
```diff
@@ -1,8 +1,8 @@
 import logging
```
Comment (wlach) [inline treeherder/autoclassify/detect_intermittents.py]: Given that we aren't running this, I feel like it might be better just to remove it rather than continuing to maintain code that we're not really testing.

File: tests/ui/unit/controllers/autoclassify.tests.js (hunk 8 above)
Comment (jgraham) [inline tests/ui/unit/controllers/autoclassify.tests.js]: Removed it for now. I think tests here probably are useful though.

File: treeherder/autoclassify/detect_intermittents.py (hunk 14 above)
Comment (jgraham) [inline treeherder/autoclassify/detect_intermittents.py]: Well we are still running the unit tests for it.

File: treeherder/model/migrations/0055_autoclassify_status.py (hunk 9 above)
Comment (jgraham) [inline treeherder/model/migrations/0055_autoclassify_status.py]: Only if the `set foreign_key_checks` thing is a good idea in this case. Maybe it's not since no foreign keys are affected.

File: treeherder/autoclassify/detect_intermittents.py (hunk 14 above)
Comment (wlach) [inline treeherder/autoclassify/detect_intermittents.py]: Yes but that's not at all the same as code we're running in production. I'm not going to block review on this, but IMO you should just remove these files as they're apt to confuse people. If it turns out that we want to resurrect this code, it's easy to go back in history.

Comment (jgraham) [inline treeherder/autoclassify/detect_intermittents.py]: I don't think it's "different" to the code we're running in production; it's just preffed off. That seems fine to me. Nevertheless, I removed it, but it at least seems pretty weird to bundle it with these unrelated changes.

File: ui/plugins/failure_summary/main.html [hunk 15]
```diff
@@ -5,7 +5,6 @@
             <div class="job-tabs-content">
                 <a class="btn btn-xs btn-default"
                    prevent-default-on-left-click
-                   ng-show="user.loggedin && (filerInAddress || user.is_staff)"
```
Comment (edmorley) [inline ui/plugins/failure_summary/main.html]: Did you mean to make this change?

File: ui/plugins/controller.js [hunk 16]
```diff
@@ -56,7 +56,8 @@ treeherder.controller('PluginCtrl', [
         // phase of this feature.
         var showAutoClassifyTab = function() {
             thTabs.tabs.autoClassification.enabled = ($location.search().autoclassify === true ||
-                                                      $location.host().indexOf('herokuapp.com') !== -1) &&
+                                                      $location.host().indexOf('herokuapp.com') !== -1 ||
```
Comment (edmorley) [inline ui/plugins/controller.js]: This could be in a separate PR.

File: ui/partials/main/thShortcutTable.html [hunk 17]
```diff
@@ -14,56 +14,86 @@
...
       -->
-      <tr><td class="kbd">esc</td>
-        <td>Close all open panels</td></tr>
-      <tr><td class="kbd">f</td>
-        <td>Enter a quick filter</td></tr>
       <tr>
-        <td><span class="kbd">ctrl</span><span class="kbd">shift
-          </span><span class="kbd">f</span></td>
-        <td>Clear the quick filter</td>
+        <th colspan=2>Jobs View</th>
```
Comment (edmorley) [inline ui/partials/main/thShortcutTable.html]: Much of the changes in this file could be in a separate PR.

File: ui/js/controllers/main.js [hunk 18]
```diff
@@ -341,14 +341,14 @@ treeherderApp.controller('MainCtrl', [
             }],
 
             // Shortcut: select next unclassified failure
-            [['j', 'n'], function() {
+            ['n', function() {
```
Comment (edmorley) [inline ui/js/controllers/main.js]: I can't speak for what the current sheriffs use, but I only ever used j/k and not n/p, so this change would have really messed with my muscle memory. Though happy for them to make the call on this.

File: treeherder/autoclassify/detect_intermittents.py (hunk 14 above)
Comment (wlach) [inline treeherder/autoclassify/detect_intermittents.py]: You could have opened another pull request for the removal, I wasn't saying that it had to happen as part of this PR.

File: ui/js/controllers/bugfiler.js [hunk 19]
```diff
@@ -2,12 +2,14 @@
...
-    'thPinboard', 'thEvents', 'fullLog', 'parsedLog', 'reftest', 'selectedJob', 'allFailures',
-    'thNotify',
+    'fullLog', 'parsedLog', 'reftest', 'selectedJob', 'allFailures',
+    'successCallback', 'thNotify',
     function BugFilerCtrl(
         $scope, $rootScope, $uibModalInstance, $http, summary, thBugzillaProductObject,
-        thPinboard, thEvents, fullLog, parsedLog, reftest, selectedJob, allFailures,
-        thNotify) {
+        fullLog, parsedLog, reftest, selectedJob, allFailures,
+        successCallback, thNotify) {
```
Comment (edmorley) [inline ui/js/controllers/bugfiler.js]: Yeah please can this be in a separate PR.

File: ui/css/treeherder-global.css [hunk 20]
```diff
@@ -333,9 +333,11 @@ input:focus::-moz-placeholder {
 }
 
 #onscreen-shortcuts {
-    position: relative;
-    top: 130px;
+    position: fixed;
```
Comment (edmorley) [inline ui/css/treeherder-global.css]: Separate PR please!

File: treeherder/model/migrations/0055_autoclassify_status.py (hunk 9 above)
Comment (edmorley) [inline treeherder/model/migrations/0055_autoclassify_status.py]: In cases where we override the default migrations, I think we should include the original sqlmigrate output in a comment in the migration file.

I'll check the migrations more thoroughly when this is added.

File: treeherder/model/migrations/0055_autoclassify_status.py [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('model', '0054_duplicate_failure_classifications'),
+    ]
+
+    operations = [migrations.RunSQL(
+        sql="""
+BEGIN;
```
Comment (edmorley) [inline treeherder/model/migrations/0055_autoclassify_status.py]: Would you mind using the list form of RunSQL rather than a string, since the latter has to be parsed by sqlparse, which I'd just managed to remove our use of (I need to update the Django docs to make the list form more obvious).

eg:
https://github.com/mozilla/treeherder/blob/5318924e00157ac8dc76351d527a970a4f6db26f/treeherder/model/migrations/0001_squashed_0053_add_job_platform_option_push_index.py#L725-L728

File: treeherder/log_parser/tasks.py [hunk 22]
```diff
@@ -88,8 +90,19 @@ def crossreference_error_lines(job_id, priority):
     newrelic.agent.add_custom_parameter("job_id", job_id)
     logger.debug("Running crossreference-error-lines for job %s" % job_id)
     job = Job.objects.get(id=job_id)
-    has_lines = crossreference_job(job)
+    try:
+        has_lines = crossreference_job(job)
+    except Exception as e:
+        import traceback
+        logger.error(traceback.format_exception(e))
```
Comment (edmorley) [inline treeherder/log_parser/tasks.py]: Wouldn't `logger.exception()` give the same result, but without the need to use `traceback` manually?

That said, do we need this? We don't do any such special handling anywhere else.

File: treeherder/log_parser/tasks.py [hunk 23]
```diff
@@ -64,6 +64,7 @@ def parse_log(job_log, priority):
     Call ArtifactBuilderCollection on the given job.
     """
     post_log_artifacts(job_log)
+    logger.debug("Scheduling crossreference for job %i from parse_log" % job_log.job.id)
```
Comment (edmorley) [inline treeherder/log_parser/tasks.py]: This could be in a separate PR.

File: treeherder/log_parser/crossreference.py [hunk 24]
```diff
@@ -63,6 +76,8 @@ def _crossreference(job):
                 summary=summary,
                 line_number=error.line_number,
                 failure_line=failure_line))
+            error.failure_line = failure_line
+            error.save()
```
Comment (edmorley) [inline treeherder/log_parser/crossreference.py]: This looks like it could be in a separate PR?

File: treeherder/autoclassify/management/commands/detect_intermittents.py [hunk 25]
```diff
@@ -1,15 +0,0 @@
-from django.core.management.base import BaseCommand
-
-from treeherder.autoclassify.detect_intermittents import detect
-from treeherder.model.models import Job
-
-
-class Command(BaseCommand):
```
Comment (edmorley) [inline treeherder/autoclassify/management/commands/detect_intermittents.py]: Could we remove the detect intermittents code in a separate, earlier PR?

**Comment by edmorley** [comment]
Just spotted this New Relic exception on treeherder-prototype:
https://rpm.newrelic.com/accounts/677903/applications/7385291/filterable_errors?tw%5Bend%5D=1487030691&tw%5Bstart%5D=1487028891#/show/bdac3cb6-f248-11e6-9c7c-b82a72d22a14_0_4582/stack_trace?top_facet=transactionUiName&primary_facet=error.class&barchart=barchart&_k=5hdtqb

KeyError during the autoclassify task (autoclassify.py line 84, in update_db).

File: treeherder/model/migrations/0055_autoclassify_status.py (hunk 21 above)
Comment (jgraham) [inline treeherder/model/migrations/0055_autoclassify_status.py]: OK, but why? This is a pretty substantial extra burden since there is no longer a string in the migration file that you can copy and check for syntax errors / run in a SQL command line. I think that the more requirements we place on people trying to use migrations for invasive changes, the more likely they are to decide to run the SQL manually against production instead.

Comment (edmorley) [inline treeherder/model/migrations/0055_autoclassify_status.py]: Ok fair point, happy to leave as is.

**Comment by edmorley** [comment]
Just thought - it's probably worth testing the "ES isn't present" codepaths still work (since on initial landing the ES instances won't exist or be populated for stage/prod). On prototype I'm presuming this could be tested by clearing the `ELASTICSEARCH_URL` environment variable temporarily.

On a similar note, do we think we need to test the "ES is present but is empty" case on prototype too? Seeing as we'll presumably need to run the backfill script on stage/prod so should test it works still.

File: ui/js/models/job_artifact.js [hunk 26]
```diff
@@ -0,0 +1,45 @@
+'use strict';
+
+treeherder.factory('ThJobArtifactModel', [
```
Comment (wlach) [inline ui/js/models/job_artifact.js]: LOL no, let's not add this monster back. ;)

File: ui/plugins/auto_classification/main.html [hunk 27]
```diff
@@ -1,181 +1,10 @@
...
+<span style="display:none"></span><!-- because otherwise the entire panel seems to be blank -->
+<th-autoclassify-panel
+   th-job="job"
+   logs-parsed="jobLogsAllParsed"
+   has-logs="job_log_urls.length > 0"
+   log-parse-status="logParseStatus"
+   autoclassify-status="job.autoclassify_status"
+   th-user="user"
+   >
+</th-autoclasify-panel>
```
Comment (wlach) [inline ui/plugins/auto_classification/main.html]:
```html
</th-autoclassify-panel>
```
